### DISCUSSION TITLE
Custom Node with Inherit<>

### DISCUSSION THREAD

**Discussion by Anshi701** [vsg-dev/VulkanSceneGraph#990] [Q&A] [answered]
Hello, i'm making a project with VSG and encouneterd with some situation. vsgExchange reading 3D model as MatrixTransform node. I wanted to make this node have some bool value to switch on\off this node (let forget about StateSwitch for a moment). 
And i failed to create some child from MatrixTransfom with new field in a way that read_cast can convert it. 
Here is exaple of my child class and function call:

```c++
class vsgSceneModel : public vsg::Inherit<vsg::MatrixTransform, vsgSceneModel>{
	vsgSceneModel() {}

	bool isEnabled = true;

	void read(vsg::Input& input) {
		vsg::MatrixTransform::read(input);
	}
	void write(vsg::Output& output) const {
		if (!isEnabled) return;
		vsg::MatrixTransform::write(output);
		
	}

}

// ***

auto node = vsg::read_cast<vsg::vsgSceneModel>(filename, VARS->GetLoadOptions());

```

I short, how can i create custom behavior and be able to cast parent node to child?

**Comment by robertosfield** [comment, marked as answer]
The read and write methods are just for reading and writing members to Input/Output, they are aren't for runtime behaviors.

The vsg::Switch node is now I would use to toggle on/off the traversal of a subgraph, so I wouldn't use custom node, it's just easier to decorate your subgraph with a vsg::Switch node and then use it's functionality to toggle on/off traversal.

If you really want to make a transform node that also acts as a switch then you could do the above but just read/write the isEnabled flag so that you have the support for reading/writing .vsgt/.vsgb files, then to implement isEnabled I'd override the virtual void accept(ConstVisitor&), virtual void accept(Visitor&) and virtual void accept(RecordTraversal&) methods.  You may be just fine overriding the accept(RecordTraversal&) one.

However, this subclassing approach isn't very flexible as it forces you start subclassing all sorts of standard scene graph nodes just to add the toggle, it's much more flexible to just decorate the nodes you want to toggle with a vsg::Switch node.

**Reply by Anshi701** [reply]
Thanks for your reply! I agree what for toggling nodes Switch is the solution. If subclassing is not the way, then if want to make some custom behavior (for example, highlight model), i need to create my own Node, override basic methods (like "accept"), assign child Node to it, do my stuff and call corresponding methods of child Node?

**Reply by robertosfield** [reply]
The first thing I would do is get the highlight effect you want working with just standard VSG features then figure out what needs to be done to encapsulate this in a more convenient form such as custom Node that decorates your scene graph.

The custom node can override the Node::accept(...) or Node::traverse(..) methods, it all depends upon what you are trying to achieving. 

I think a highlight node example would be a nice addition to vsgExamples, I don't have time right now to tackle this, so if anyone else fancies it then it would be a good example to others even if it doesn't quite work yet.  Such an example could evolve into a cleaner form with the help of others chipping in.

**Reply by Anshi701** [reply]
I got it, thank you!
